src/Lean/Elab/Tactic/Monotonicity.lean

```diff
@@ -184,7 +184,10 @@ def solveMonoStep (failK : ∀ {α}, Expr → Array Name → MetaM α := @defaul
     -- A recursive call
     if let some hmono ← solveMonoCall α inst_α e then
       trace[Elab.Tactic.monotonicity] "Found recursive call {e}:{indentExpr hmono}"
-      unless ← goal.checkedAssign hmono do
+      let goalAssigned ← goal.checkedAssign hmono
+      let hmonoType ← inferType hmono
+      let agreeOnType ← isDefEq hmonoType type
+      unless goalAssigned && agreeOnType do
         trace[Elab.Tactic.monotonicity] "Failed to assign {hmono} : {← inferType hmono} to goal"
         failK f #[]
```

Comment:
I think you can check the type, and if it works use `.assign`. At that point we don’t need to do anything `checked` I believe
```suggestion
      unless (← isDefEq hmonoType) do
        trace[Elab.Tactic.monotonicity] "Failed to assign {hmono} : {← inferType hmono} to goal"
        failK f #[]
      goal.assign hmono
```